CRM/Contact/BAO/Query.php

```diff
@@ -436,6 +442,11 @@ public function __construct(
       $this->_params = array();
     }
 
+    // Search by Primary location type ONLY if searchPrimaryDetailsOnly setting is on OR
+    //  search query is trigerred via api OR
+    //  special parameter search_by_primary_details_only = TRUE
+    self::$_alwaysSearchByPrimaryOnly = (Civi::settings()->get('searchPrimaryDetailsOnly') || $apiEntity || !empty($params['search_by_primary_details_only']));
```

Comment:
I think the api should set $params['search_by_primary_details_only'] rather than $apiEntity being code for that.